spark_log_parser/loaders/__init__.py

```diff
@@ -61,58 +71,110 @@ def __iter__(self):
+
+            # If we have enough data in the buffer to satisfy this read, return that from the buffer and adjust
+            #  our read position accordingly
+            if self._num_bytes_left_in_buffer() >= size:
+                start_pos = self._buffer_position
+                self._buffer_position += size
+                return self._buffer_memview[start_pos:self._buffer_position]
+            else:
+                # Otherwise, we don't have enough data, which means that this is the last read for this iterator.
+                #  So now just return a copy of the data remaining in the buffer and mark this as consumed to prevent
+                #  future reads
+                content = self._buffer_memview[self._buffer_position:]
+                self._consumed = True
+                return content
+
+    _LINE_CHUNK_SIZE = 1024 * 1024
 
     def iter_lines(self) -> Iterator[bytes]:
         """
         Returns a stream of lines from self._chunks.
         """
-        for chunk in self._chunks:
-            self._add_chunk_to_size(chunk)
-            if self._trailing_data:
-                chunk = self._trailing_data + chunk
-
-            lines = chunk.splitlines()
+        trailing_data = bytearray()
+        # We use a relatively large chunk_size here to minimize the number of times we have to enter this loop/copy data
+        while chunk := self.read(self._LINE_CHUNK_SIZE):
+            if trailing_data:
+                # Instead of appending via e.g. chunk = trailing_data + chunk, which causes a ton of data copying,
+                #  we can just write this chunk into our current line buffer. This is extremely helpful when we
+                #  encounter large files where all the data is actually on one line (i.e. a 500MB JSON file that was
+                #  written without any line breaks). This saves us 10s of seconds when processing these types of files
+                # This is safe to do here because trailing_data is never referenced outside of this function, and we
+                #  only ever return trailing_data when we are done reading in the whole file
+                trailing_data.extend(chunk)
+                chunk = trailing_data
+                trailing_data = bytearray()
+
+            # In the case where we have no trailing_data, then our `chunk` may be a memoryview returned from
+            #  self.read(), which does not support splitlines(), so we need to copy this into a bytearray first
+            bs = chunk if isinstance(chunk, bytearray) else bytearray(chunk)
+
+            # Don't check for `\r\n` because that will be caught by checking for the individual characters
+            # Doing this similarly helps keep parsing fast for large single-line files, while having minimal performance
+            #  impact on files that actually have more than 1 line
+            if b'\n' in bs or b'\r' in bs:
+                lines = bs.splitlines()
+            else:
+                lines = [bs]
```

Comment:
I saw a similarly pretty wild impact on the log mentioned just above - using `in` to check for these characters prior to calling `splitlines()` saves us _over a minute_ when processing that log. This is because `splitlines()` will end up copying the _whole array_ into the returned list. This `in` check adds almost no real overhead for files that have actual lines in them, and saves us a ton of data copying for files with few/none at all. Again comparing the old implementation to the new one, we see that not only is the self-time in `self.iter_lines()` faster, but `splitlines()` doesn't even show up _at all_, for a savings of 47 + 17 = _64 seconds_ -


<img width="1487" alt="Screenshot 2023-01-24 at 1 57 16 PM" src="https://user-images.githubusercontent.com/117235708/214430139-97fdb8c7-772c-4481-84e4-87d0bac4e03b.png">

(note that the time spend in the SSL read is subject to my internet speed - the CPU time of these functions is _not_ impacted by that, however)

Reply:
Nice! Is this a screenshot of PyCharm?

Reply:
Yep! I captured these profiles from the command line, but PyCharm has a built-in `cprof` visualizer, which is pretty nice (including the Call Graph!)